File: FtClient/src/StatisticsReport.cpp
/**
 * @file StatisticsReport.hpp
 * @brief Statistics report implementation
 */

#include <chrono>
#include <iostream>
#include <iomanip>
#include "StatisticsReport.hpp"

using namespace std;

/////////////////////////////////////////////////////////////////////////////

BEGIN_NAMESPACE;

static constexpr int64_t oneSecondInUs = 1000 * 1000;

StatisticsReport::StatisticsReport()
	: lastReportTime_(chrono::high_resolution_clock::now()),
	  totalSize_(0)
{

}

void StatisticsReport::reset()
{

	this->lastReportTime_ = chrono::high_resolution_clock::now();
	this->totalSize_ = 0;
}

void StatisticsReport::addSize(uint64_t size)
{

	this->totalSize_ += size;

	const auto now = chrono::high_resolution_clock::now();
	const auto diffTime = now - this->lastReportTime_;
	const auto diffTimeUs =
	    chrono::duration_cast<chrono::microseconds>(diffTime).count();

	if (diffTimeUs > oneSecondInUs) {

		const auto megaBytesPerSecond =
		    static_cast<double>(this->totalSize_) / diffTimeUs;

		cout << fixed << setprecision(2) << megaBytesPerSecond
		     << " (MB/s) - "
		     << this->totalSize_ << " bytes in "
		     << diffTimeUs << " (us)\n";

		this->reset();
	}
}

END_NAMESPACE;

File: FtClientDotNet/Hglee.Device.Ftd3xx/StatisticsReport.cs
namespace Hglee.Device.Ftd3xx;

using System;

/// <summary>
/// Statistics report class
/// </summary>
public sealed class StatisticsReport
{
    /// <summary>
    /// 1 second in microseconds
    /// </summary>
    private static readonly double OneSecondInUs = 1000 * 1000;

    /// <summary>
    /// Last report timestamp
    /// </summary>
    private DateTime lastReportTime;

    /// <summary>
    /// Total size
    /// </summary>
    private ulong totalSize;

    /// <summary>
    /// Initializes a new instance of the <see cref="StatisticsReport"/> class.
    /// </summary>
    public StatisticsReport()
    {
        this.Reset();
    }

    /// <summary>
    /// Reset state.
    /// </summary>
    public void Reset()
    {
        this.lastReportTime = DateTime.Now;
        this.totalSize = 0;
    }

    /// <summary>
    /// Add read/write size and print statistics.
    /// </summary>
    public void AddSize(ulong size)
    {
        this.totalSize += size;

        var now = DateTime.Now;
        var diffTime = now - this.lastReportTime;
        var totalUs = diffTime.TotalMilliseconds * 1000.0;

        if (totalUs > OneSecondInUs)
        {
            var megaBytesPerSecond = (double)this.totalSize / totalUs;

            Console.WriteLine($"{megaBytesPerSecond:0.00} (MB/s) - {this.totalSize} bytes in {totalUs:0.00} (us)");

            this.Reset();
        }
    }
}
